skxray/cdi.py

```diff
@@ -119,9 +125,339 @@ def convolution(array1, array2):
+    index_v : array
+        index to define sample area
+
+    Returns
+    -------
+    sample_obj : array
+        sample object with proper cut.
+    """
+    sample_obj = np.array(sample_obj)
+    sample_obj[index_v] = 0
+    return sample_obj
+
+
+def cal_relative_error(x_old, x_new):
+    """
+    Relative error is calculated as the ratio of the difference between the new and
+    the original arrays to the norm of the original array.
+
+    Parameters
+    ----------
+    x_old : array
+        previous data set
+    x_new : array
+        new data set
+
+    Returns
+    -------
+    float :
+        relative error
+    """
+    return np.linalg.norm(x_new - x_old)/np.linalg.norm(x_old)
+
+
+def cal_diff_error(sample_obj, diff_array):
+    """
+    Calculate the error in q space.
+
+    Parameters
+    ----------
+    sample_obj : array
+        sample data
+    diff_array : array
+        diffraction pattern
+
+    Returns
+    -------
+    float :
+        relative error in q space
+    """
+    new_diff = np.abs(_fft_helper(sample_obj)) / np.sqrt(np.size(sample_obj))
+    return cal_relative_error(diff_array, new_diff)
+
+
+def generate_random_phase_field(diff_array):
+    """
+    Initiate random phase.
+
+    Parameters
+    ----------
+    diff_array : array
+        diffraction pattern
+
+    Returns
+    -------
+    sample_obj : array
+        sample information with phase
+    """
+    pha_tmp = np.random.uniform(0, 2*np.pi, diff_array.shape)
+    sample_obj = (_ifft_helper(diff_array * np.exp(1j*pha_tmp))
+                  * np.sqrt(np.size(diff_array)))
+    return sample_obj
+
+
+def generate_box_support(sup_radius, shape_v):
+    """
+    Generate support area as a box for either 2D or 3D cases.
+
+    Parameters
+    ----------
+    sup_radius : float
+        radius of support
+    shape_v : list
+        shape of diffraction pattern, which can be either 2D or 3D case.
+
+    Returns
+    -------
+    sup : array
+        support with a box area
+    """
+    sup = np.zeros(shape_v)
+    if len(shape_v) == 2:
+        nx, ny = shape_v
+        sup[nx/2-sup_radius: nx/2+sup_radius,
+            ny/2-sup_radius: ny/2+sup_radius] = 1
+    elif len(shape_v) == 3:
+        nx, ny, nz = shape_v
+        sup[nx/2-sup_radius: nx/2+sup_radius,
+            ny/2-sup_radius: ny/2+sup_radius,
+            nz/2-sup_radius: nz/2+sup_radius] = 1
+    return sup
+
+
+def generate_disk_support(sup_radius, shape_v):
+    """
+    Generate support area as a disk for either 2D or 3D cases.
+
+    Parameters
+    ----------
+    sup_radius : float
+        radius of support
+    shape_v : list
+        shape of diffraction pattern, which can be either 2D or 3D case.
+
+    Returns
+    -------
+    sup : array
+        support with a disk area
+    """
+    sup = np.zeros(shape_v)
+    dummy = _dist(shape_v)
+    sup_index = np.where(dummy <= sup_radius)
+    sup[sup_index] = 1
+    return sup
+
+
+def cdi_recon(diff_array, sample_obj, sup,
```

Comment:
We always assume the center is q=0. Finding center is another step beyond scope of the current code.